### PR TITLE
Change minimum Ruby version for Puma to 2.2 + specify rdoc version under 2.2

### LINKED ISSUE #2012: [CI] Deal with Puma requiring ruby 2.2+ minimum
Hello

Recently [Puma added a limitation on Ruby 2.2+](https://github.com/puma/puma/pull/1506). So our CI is breaking now (https://travis-ci.org/rspec/rspec-rails/jobs/404060429)

capture d ecran 2018-07-15 a 11 01 18

I'm wondering how can we handle this? If the Ruby version is below 2.2 we use an old Puma version?

### PR DESCRIPTION
**Pull request by benoittgt**
Fix #2012

Puma now requires Ruby 2.2 minimum https://github.com/puma/puma/pull/1506/files

### CODE REVIEW

**Comment by benoittgt** [comment]
The issue is now on `rdoc`
```
Using rails 4.0.13
Installing rdoc 6.0.4
Gem::RuntimeRequirementNotMetError: rdoc requires Ruby version >= 2.2.2. The
current ruby version is 2.1.0.
An error occurred while installing rdoc (6.0.4), and Bundler cannot
continue.
Make sure that `gem install rdoc -v '6.0.4' --source 'https://rubygems.org/'`
```

Because of https://github.com/ruby/rdoc/commit/51b03c357457183e94a6b26d71630f183afbc6b0

**Comment by benoittgt** [comment]
It doesn't seem this is the way to fix rdoc issue. https://travis-ci.org/rspec/rspec-rails/builds/404100674

Do I miss something?

**Comment by benoittgt** [comment]
It seems to be better now. Even if it's not very stable (especially on rails 3).

Feel free to merge it if it's ok RSpec core team members.

File: Gemfile [hunk 1]
```diff
@@ -24,6 +24,11 @@ end
 
 gem 'sqlite3', '~> 1.3.6'
 
+# Version 6.0.4 of rdoc requires Ruby 2.2
+if RUBY_VERSION < '2.2'
+  gem 'rdoc', '< 6.0.4'
+end
```
Comment (JonRowe) [inline Gemfile]: If rdoc is failing it's likely down in the app installs, not in our outer Gemfile, you can see in the example app generator gsub rules for generated Gemfiles which help to mitigate this.

File: Gemfile-rails-dependencies [hunk 2]
```diff
@@ -13,14 +13,14 @@ when /master/
...
-  if RUBY_VERSION  >= "1.9.3"
+  if RUBY_VERSION  >= "2.2"
     gem 'puma', :git => 'git://github.com/puma/puma', :branch => 'master'
   end
 when /stable$/
   gem_list = %w[rails railties actionmailer actionpack activerecord activesupport]
   gem_list << 'activejob'  if version > '4-1-stable'
   gem_list << 'actionview' if version > '4-0-stable'
-  if RUBY_VERSION  >= "1.9.3"
+  if RUBY_VERSION  >= "2.2"
```
Comment (JonRowe) [inline Gemfile-rails-dependencies]: So we'd normally do a gate here, install an older version of puma on platforms that support it, and install the latest on Ruby 2.2, this is necessary to support older rubies.

File: Gemfile-rails-dependencies [hunk 3]
```diff
@@ -40,7 +40,7 @@ when nil, false, ""
 else
   gem "rails", version
 
-  if version >= '5-1-stable' && RUBY_VERSION  >= "1.9.3"
+  if version >= '5-1-stable' && RUBY_VERSION  >= "2.2"
```
Comment (JonRowe) [inline Gemfile-rails-dependencies]: As above.

**Comment by benoittgt** [comment]
This CI is a nightmare. I tried only on failing build on my machine and it's working. I hope I understand properly your comments @JonRowe. I will look at it tonight (UTC) again.

File: example_app_generator/generate_app.rb [hunk 4]
```diff
@@ -47,6 +47,11 @@
     |  gem 'mime-types', '< 3'
     |end
     |
+    |# Version 6.0.4 of rdoc requires Ruby 2.2
+    |if RUBY_VERSION < '2.2'
+    |  gem 'rdoc', '< 6.0.4'
```
Comment (JonRowe) [inline example_app_generator/generate_app.rb]: It's better to do a gsub, to specify the version only if rails adds it, and ensures theres only one, see line 24

File: Gemfile-rails-dependencies [hunk 5]
```diff
@@ -13,14 +13,16 @@ when /master/
...
+  if RUBY_VERSION  >= "2.2"
     gem 'puma', :git => 'git://github.com/puma/puma', :branch => 'master'
   end
 when /stable$/
   gem_list = %w[rails railties actionmailer actionpack activerecord activesupport]
   gem_list << 'activejob'  if version > '4-1-stable'
   gem_list << 'actionview' if version > '4-0-stable'
-  if RUBY_VERSION  >= "1.9.3"
+  if RUBY_VERSION  >= "1.9.3" && RUBY_VERSION < "2.2"
+    gem "puma", "~> 3.11.4"
```
Comment (JonRowe) [inline Gemfile-rails-dependencies]: One of these (this line, and line 26) is wrong, but I'm not sure which, can puma ever be found in the rails repo?

**Comment by benoittgt** [comment]
I read this https://github.com/rspec/rspec-rails/commit/087cb16b3012ec47e7a4a83e192f2495b38057ab and this https://github.com/rspec/rspec-rails/commit/aa93e588e5be3c65c22290ba3bfb3cbfa24865d5 and it helps me to understand what could I do.

For me we don't need puma before Ruby 2.2. Because it seems [we need puma for system test](https://github.com/rspec/rspec-rails/commit/2fb410b2670fdaf22f6bd2d0b8d67f9d1b78334d ) added into Rails 5.1. The minimum requirement for Rails 5.1 is Ruby 2.2 https://github.com/rails/rails/blob/5-1-stable/rails.gemspec#L10

So maybe the initial commit is not a bad idea https://github.com/rspec/rspec-rails/pull/2013/commits/33e41f254eb8596165edd94e866075279ed73925

For rdoc you did the change few week ago. https://github.com/rspec/rspec-rails/commit/6a3e2bce4f252bd50aae3deadc1067ec12a8727a

So I rebase from master and keep only my initial commit. We will see what CI say.

**Comment by JonRowe** [comment]
Actually even Rails 5 requires < Ruby 2.2 so this should be safe

**Comment by JonRowe** [comment]
If this goes green merge it @benoittgt, sorry for the wild goose chase!

**Comment by benoittgt** [comment]
The error is now somewhere else

```
bundle exec rails new ./tmp/example_app --no-rc --skip-javascript --skip-sprockets --skip-git --skip-test-unit --skip-listen --skip-bundle --template=example_app_generator/generate_app.rb
Traceback (most recent call last):
	3: from /home/travis/.rvm/gems/ruby-2.5.1/bin/ruby_executable_hooks:24:in `<main>'
	2: from /home/travis/.rvm/gems/ruby-2.5.1/bin/ruby_executable_hooks:24:in `eval'
	1: from /home/travis/.rvm/gems/ruby-2.5.1/bin/bundle:23:in `<main>'
/home/travis/.rvm/gems/ruby-2.5.1/bin/bundle:23:in `load': cannot load such file -- /home/travis/.rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/bundler-1.16.3/exe/bundle (LoadError)
rake aborted!
Command failed with status (1): [bundle exec rails new ./tmp/example_app --...]
/home/travis/build/rspec/bundle/ruby/2.5.0/gems/rake-12.3.1/lib/rake/file_utils.rb:67:in `block in create_shell_runner'
/home/travis/build/rspec/bundle/ruby/2.5.0/gems/rake-12.3.1/lib/rake/file_utils.rb:57:in `sh'
/home/travis/build/rspec/rspec-rails/Rakefile:63:in `block (2 levels) in <top (required)>'
```

I'm gonna take a look

**Comment by benoittgt** [comment]
Ok. This seems linked https://github.com/bundler/bundler/issues/6629

**Comment by benoittgt** [comment]
I will monitor the status of this issue on bundler and will remove the specific version of bundler when it will be fixed.

**Pull request merged**